app/src/main/java/com/github/polybooks/core/database/implementation/OLBookDatabase.kt

```diff
@@ -0,0 +1,259 @@
+private const val ISBN_FORMAT = """($ISBN10_FORMAT)|($ISBN13_FORMAT)"""
+
+private const val OL_BASE_ADDR = """https://openlibrary.org"""
+
+/**
+ * An implementation of a book database based on the Open Library online database
+ * */
+class OLBookDatabase(private val url2json : (String) -> CompletableFuture<JsonElement>) : BookDatabase {
+
+    override fun queryBooks(): BookQuery = OLBookQuery()
+
+    private inner class OLBookQuery : BookQuery {
+
+        private var ordering = DEFAULT
+
+        private var empty: Boolean = true
+        private var title: String? = null
+        private var isbn: String? = null
+
+        override fun onlyIncludeInterests(interests: Collection<Interest>): BookQuery {
+            System.err.println("Warning: onlyIncludeInterest not fully implemented for OLBookQuery")
+            this.empty = true
+            return this
+        }
+
+        override fun searchByTitle(title: String): BookQuery {
+            System.err.println("Warning: search by title not fully implemented for OLBookQuery")
+            this.empty = true
+            return this
+        }
+
+        override fun searchByISBN13(isbn13: String): BookQuery {
+            val regularised = regulariseISBN(isbn13) ?: throw IllegalArgumentException("Given ISBN is not valid")
+            this.empty = false
+            this.title = null
+            this.isbn = regularised
+            return this
+        }
+
+        override fun withOrdering(ordering: BookOrdering): BookQuery {
+            this.ordering = ordering
+            return this
+        }
+
+        @RequiresApi(Build.VERSION_CODES.N)
+        override fun getAll(): CompletableFuture<List<Book>> {
+            if (empty) return CompletableFuture.completedFuture(Collections.emptyList())
+            else {
+                assert(isbn != null)
+                val url = isbn2URL(isbn!!)
+                return url2json(url)
+                        .thenApply { parseBook(it) }
+                        .thenCompose { updateBookWithAuthorName(it) }
+                        .thenApply { listOf(it) }
+                        .exceptionally { exception ->
+                            if (exception is CompletionException && exception.cause is FileNotFoundException)
+                                return@exceptionally Collections.emptyList<Book>()
+                            else if (exception is CompletionException) throw exception.cause!!
+                            else throw exception
+                        }
+            }
+        }
+
+        @RequiresApi(Build.VERSION_CODES.N)
+        override fun getN(n: Int, page: Int): CompletableFuture<List<Book>> {
+            if (n < 0 || page < 0) {
+                throw IllegalArgumentException(
+                    if (n < 0) "Cannot return a negative ($n) number of results"
+                    else "Cannot return a negative ($page) page number"
+                )
+            }
+            return getAll()
```

Comment:
Why return `getAll` and ignore `n` and `page`? Is it meant to be implemented later?

Reply:
because for now, A OLBookQuery can only get one book, so the pages don't really matter